### ISSUE TITLE
Use a "one folder per notebook" layout

### ISSUE THREAD

**Issue by eteq** [adrn/nbstatic#1]
This is content from @jbcurtin that came up in an out-of-band conversation that I'm moving here so @adrn (and any others) can see:

To reduce confusion in communication, I've been holding back on version types between the three repos

* spacetelescope/jdat_notebooks - most up to date
* spacetelescope/notebooks         - initial implementation
* spacetelescope/py_datinthesky - discussion needed
 

jdat_notebooks has the latest code version. I'll be upgrading notebooks next week with the most up to date from jdat_notebooks. To upgrade py_datinthesky, I recommend we discuss what the folder structure layout. I like the layout of nbstatic, but it assumes you can have two notebooks in one folder. I'd like to recommend against this as it could be an anti-pattern when trying to organize files. For example, organizing 100+ notebooks in a single folder could easily become a maintenance nightmare

Here are the reasons why I think having all notebooks in one folder could become a maintenance nightmare

* Too much changes during build
* Each notebook will at least have one requirements.txt file, [ Notebook Name ].ipynb file, and environment.sh file
* Each notebook will probably download multiple, if not hundreds of fits or asdf files
* This could easily turn into a build challenge too
* Having to remove 25 notebooks to focus an a single notebook would be an extract step when having to debug
* Scientists would have to check to see if the folder has all the configuration files required to check into git
* Each notebook may have a pre_requirements.txt file to augment the requirements.txt file along with a pre_install.sh script to prepare the build pipeline and install things like gcc or something else. The idea of packaging everything in conda works 90% of the time, however we have to consider the 10% off chance of something poping up...which it already has in jdat_notebooks
Some of the dependencies in 'notebooks' and 'jdat_notebooks' have required numpy to be installed before something else can be installed
 

Influenced by notebooks, nbpages, and nbstatic, The notebook layout I'd recommend is,

 
```
spacetelescope/jdat_notebooks
-> notebooks
---> [ Notebook Name ]
-------> [ Notebook Name ].ipynb
-------> [ some configuration ].(toml|cfg)
-------> pre_install.sh
-------> pre_requirements.sh
-------> requirements.txt

---> [ Notebook Name ]
-------> [ Notebook Name ].ipynb
-------> requirements.txt
```

It could further be optimized into,
 

 
```
spacetelescope/jdat_notebooks
-> notebooks
---> [ Notebook Name ]
------> [ Notebook Name ].ipynb
------> [ some configuration ].(toml|cfg)
------> build-tools
--------> pre_install.sh
--------> pre_requirements.sh
--------> requirements.txt

---> [ Notebook Name ]
-----> [ Notebook Name ].ipynb
-----> requirements.txt
 ```

After the html artifacts are built, we can move the html anywhere and align the rendering style of astropy to something more human friendly.

**Comment by adrn** [comment]
(FYI: `nbstatic` doesn't require having all notebooks in one folder.)

**Comment by eteq** [comment]
@adrn - right, but it's *possible* to do so, right?  

I think @jbcurtin's point is that we want to *require* notebooks to be one-notebook-per-folder.  And I agree - I agree with @jbcurtin's points and in particular IMHO I think it gets very confusing when downloaded/saved/etc files get jumpled together between multiple notebooks.

A middle ground is for that to be the *encouraged* layout, without being the *required* layout.  But then I don't really know how to implement things like ``requirements.txt`` which we know we want to be notebook-specific.

**Comment by adrn** [comment]
I don't think we should impose this on users, but encouraging good organization sounds fine!

I'm still not sold on the "config and requirements per notebook" strategy (not that I am sold on, or love, the global solution either!) - I think there are pros and cons to both options, but have been hoping a 3rd, better option would crop up somewhere :). 

**Config and requirements per notebook:**
Pros: It's clear what notebooks require what, it's faster to run tests or build a single notebook.
Cons: Big maintenance burden for repos with many notebooks, would want extra code to compile all individual dependencies into a global dependency folder

**Global requirements and config**:
Pros: Dependencies appear in one place
Cons: Running tests or building a single notebook requires installing all dependencies, less flexible (if different notebooks have different requirements - but IMO I don't think we should have to support that case because it's confusing...).

**Comment by jbcurtin** [comment]
Okay, great. I think we have a path forward here. From our discussion, here are the changes I'll be implementing

  With enabling multiple notebooks to share the same `requirements.txt` file in a single folder. It only alters `sub_path` into `[ Notebook Category ]`.

  `requirements.txt`, `install.sh`, and `pre_install.sh` are all still required by notebooks in `spacetelescope/jdat_notebooks`. `build-tools` was introduced to organize the files required for running builds. Here is the planned directory structure without having the `build-tools` folder anymore

  `[ some configuration ].(toml|cfg)` was introduced as something that maybe seen in the future. `spacetelescope/jdat_notebooks` and other repos don't yet have a need for it. I'll remove it from the planned directory structure

```
adrn/nbcollection
-> notebooks
---> [ Notebook Category ]
------> [ Notebook Name-1 ].ipynb
------> [ Notebook Name-2 ].ipynb
------> [ Notebook Name-3 ].ipynb
------> pre_install.sh
------> pre_requirements.sh
------> requirements.txt
```

Original Folder Structure
https://github.com/adrn/nbcollection#converting-a-directory-structure-of-specific-notebook-files
